DiscordBot/Program.cs

```diff
@@ -40,9 +47,83 @@ private static async Task MainAsync()
             commands.RegisterCommands<EventCommands>();
 
             discord.UseInteractivity(new InteractivityConfiguration());
+            discord.MessageReactionAdded += async (client, eventArguments) =>
+            {
+                await OnMessageReactionAdded(client, eventArguments, eventsSheetsService);
+            };
 
             await discord.ConnectAsync();
             await Task.Delay(-1);
         }
+
+        private static async Task OnMessageReactionAdded(
+            DiscordClient client,
+            MessageReactionAddEventArgs eventArguments,
+            IEventsSheetsService eventsSheetsService)
+        {
+            if (eventArguments.User == client.CurrentUser)
+            {
+                return;
+            }
+
+            var discordEvent = await GetEventFromMessageIdOrDefaultAsync(eventArguments.Message.Id, eventsSheetsService);
+            if (discordEvent == null)
+            {
+                return;
+            }
+
+            await ProcessReaction(client, eventArguments, discordEvent, eventsSheetsService);
+
+            await UpdateSignupSheet(eventArguments, discordEvent, eventsSheetsService);
+        }
+
+        private static async Task ProcessReaction(
+            DiscordClient client,
+            MessageReactionAddEventArgs eventArguments,
+            DiscordEvent discordEvent,
+            IEventsSheetsService eventsSheetsService)
+        {
+            var dmChannel = await eventArguments.Guild
+                .GetMemberAsync(eventArguments.User.Id).Result
+                .CreateDmChannelAsync();
+
+            if (eventArguments.Emoji.GetDiscordName() == ":no_entry_sign:")
+            {
+                await eventsSheetsService.ClearResponsesForUserAsync(discordEvent.Key, eventArguments.User.Id);
+                await client.SendMessageAsync(dmChannel, $"You've signed off {discordEvent.Name}.");
+            }
+            else
+            {
+                try
+                {
+                    await eventsSheetsService.AddResponseForUserAsync(
+                        discordEvent.Key,
+                        eventArguments.User.Id,
+                        eventArguments.Emoji.GetDiscordName());
+                }
+                catch (ResponseNotFoundException)
+                {
+                    return;
+                }
+
+                await client.SendMessageAsync(
```

Comment:
This is incorrect in the response not found case

Reply:
I think the method should return if the response is not found?

Reply:
Yes, good point 🤦 Might be useful to print an error to the user in that case? 🙂

Reply:
Any valid reaction will be automatically added by the bot so for a user to give an invalid reaction they will have had to manually choose an invalid one. The invalid reaction will still be removed by the bot so other users shouldn't mistake it for a valid reaction.

Reply:
Okay, sounds fine then 🙂